File: apps/swift-ios/Core/ToolActivity.swift
import Foundation

public struct ToolActivityNativeAppReference: Codable, Equatable, Sendable {
    public let _tag: String
    public let appId: String?
    public let displayName: String?
}

public struct ToolActivityIcon: Codable, Equatable, Sendable {
    public let _tag: String
    public let pageUrl: String?
    public let faviconUrl: String?
    public let faviconUrlDark: String?
    public let app: ToolActivityNativeAppReference?
    public let logoUrl: String?
    public let logoUrlDark: String?

    public func imageURL(dark: Bool) -> URL? {
        func image(_ raw: String?) -> URL? {
            guard let raw, raw.count <= 4096, let url = URL(string: raw),
                ["http", "https"].contains(url.scheme?.lowercased() ?? "") else { return nil }
            return url
        }
        switch _tag {
        case "themed-logo": return image(dark ? logoUrlDark ?? logoUrl : logoUrl)
        case "website":
            if dark, let explicit = image(faviconUrlDark) { return explicit }
            if dark, let page = image(pageUrl), page.host?.lowercased() == "github.com" {
                return URL(string: "https://github.githubassets.com/favicons/favicon-dark.svg")
            }
            if let explicit = image(faviconUrl) { return explicit }
            if let page = image(pageUrl), page.host?.lowercased() == "github.com" {
                return URL(string: "https://github.githubassets.com/favicons/favicon.svg")
            }
            guard let page = image(pageUrl), var parts = URLComponents(url: page, resolvingAgainstBaseURL: false) else { return nil }
            parts.path = "/favicon.ico"; parts.query = nil; parts.fragment = nil; parts.user = nil; parts.password = nil
            return parts.url
        default: return nil
        }
    }
}

public struct ToolActivitySource: Codable, Equatable, Sendable {
    public let key: String
    public let name: String
    public let kind: String
    public let icon: ToolActivityIcon?
}

File: packages/contracts/src/toolActivity.ts
import * as Schema from "effect/Schema";
import { TrimmedNonEmptyString as TrimmedNonEmptyStringSchema } from "./baseSchemas.ts";

export const ToolActivitySurface = Schema.Literals(["browser", "computer"]);
export type ToolActivitySurface = typeof ToolActivitySurface.Type;

export const ToolActivityNativeAppReference = Schema.Union([
  Schema.TaggedStruct("app-id", {
    appId: TrimmedNonEmptyStringSchema.check(
      Schema.isMaxLength(512),
      Schema.isPattern(/^[A-Za-z0-9._-]+$/u),
    ),
  }),
  Schema.TaggedStruct("display-name", {
    displayName: TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(160)),
  }),
]);
export type ToolActivityNativeAppReference = typeof ToolActivityNativeAppReference.Type;

export const ToolActivityIcon = Schema.Union([
  Schema.TaggedStruct("website", {
    pageUrl: TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(4096)),
    faviconUrl: Schema.optional(TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(4096))),
    faviconUrlDark: Schema.optional(TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(4096))),
  }),
  Schema.TaggedStruct("native-app", {
    app: ToolActivityNativeAppReference,
  }),
  Schema.TaggedStruct("themed-logo", {
    logoUrl: TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(4096)),
    logoUrlDark: Schema.optional(TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(4096))),
  }),
]);
export type ToolActivityIcon = typeof ToolActivityIcon.Type;

export const ToolActivitySource = Schema.Struct({
  key: TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(512)),
  name: TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(160)),
  kind: Schema.Literals(["browser", "computer", "integration"]),
  icon: Schema.optional(ToolActivityIcon),
});
export type ToolActivitySource = typeof ToolActivitySource.Type;
